### PR TITLE
refactor(rpc): change the  get balance interface

### PR DESCRIPTION
**Pull request by KaoImin**
**What this PR does / why we need it**:
1. Change the argument of the `get_balance` interface to a struct.
2. Change the `udt_hashes` to a `HashSet`, if it is empty, means to get the balance of all assets.

**Which issue(s) this PR fixes**:

Fixes #

**Which docs this PR relation**:

Ref #

**Which toolchain this PR adaption**:

No Breaking Change

**Special notes for your reviewer**:

### CODE REVIEW

File: core/rpc/src/rpc_impl.rs [hunk 1]
```diff
@@ -162,6 +166,31 @@ pub fn address_to_script(payload: &AddressPayload) -> packed::Script {
...
+    } else {
+        Err(MercuryError::rpc(RpcError::KeyAddressIsNotSecp256k1).into())
+    }
+}
+
+pub fn parse_normal_address(addr: &str) -> Result<Address> {
+    Address::from_str(addr).map_err(|e| anyhow!("{:?}", e))
+}
+
+pub fn lock_args_to_sp_key(lock_args: Bytes) -> H160 {
```
Comment (rev-chaos) [inline core/rpc/src/rpc_impl.rs:184]: This function is used to convert pubkey_hash to Secp short address, so I suggest to rename `pubkey_to_secp_address`, which can be used for either acp or cheque.

Comment (KaoImin) [reply core/rpc/src/rpc_impl.rs:184]: The secp address length is not 20, this is hashed by blake2b160.

File: core/rpc/src/error.rs [hunk 2]
```diff
@@ -56,4 +56,10 @@ pub(crate) enum RpcError {
 
     #[display(fmt = "Can not find change cell")]
     CannotFindChangeCell,
+
+    #[display(fmt = "Key address must be an Secp256k1 address")]
+    KeyAddressIsNotSecp256k1,
+
+    #[display(fmt = "Invalid normal address")]
+    InvalidNormalAddress,
```
Comment (rev-chaos) [inline core/rpc/src/error.rs:64]: Suggest renaming as `UnsupportNormalAddress`

File: core/rpc/src/rpc_impl/query.rs [hunk 3]
```diff
@@ -26,25 +32,132 @@ where
...
+                if self.is_secp256k1(&script) {
+                    self.inner_get_secp_balance(udt_hashes, &addr)?
+                } else if self.is_acp(&script) {
+                    self.inner_get_acp_balance(udt_hashes, &addr)?
+                } else {
+                    return Err(MercuryError::rpc(RpcError::InvalidNormalAddress).into());
+                }
+            }
+        };
+
```
Comment (rev-chaos) [inline core/rpc/src/rpc_impl/query.rs:80]:
```suggestion
// todo: After support search according to height, the process should be:  if search for latest height (default mode), we should get `CURRENT_BLOCK_NUMBER` first, and then query according to height
```

File: core/rpc/src/rpc_impl/query.rs [hunk 4]
```diff
@@ -26,25 +32,132 @@ where
 {
     pub(crate) fn inner_get_balance(
         &self,
-        udt_hashes: Vec<Option<H256>>,
+        udt_hashes: HashSet<Option<H256>>,
+        address: QueryAddress,
+        block_number: Option<u64>,
```
Comment (rev-chaos) [inline core/rpc/src/rpc_impl/query.rs:37]: // Not support yet, so we should return error, when the block_number is `none`, which is missed now

File: core/rpc/src/rpc_impl/query.rs [hunk 5]
```diff
@@ -26,25 +32,132 @@ where
...
-                fleeting,
-                locked,
-            ));
+            let unconstrained = if let Some(hash) = hash.clone() {
+                self.udt_balance(addr, &hash)?.unwrap_or(0)
+            } else {
+                self.ckb_balance(addr)? as u128
+            };
+
+            balances.push(Balance::new(addr.to_string(), hash, unconstrained, 0, 0));
```
Comment (rev-chaos) [inline core/rpc/src/rpc_impl/query.rs:99]: address here should be key_address.

**Review by rev-chaos — APPROVED** [review]

**Pull request merged**